### DISCUSSION TITLE
Reliable way to check if current thread is the main thread in console and WPF applications

### DISCUSSION THREAD

**Discussion by alexey-leonovich** [dotnet/runtime#110141] [Q&A]
I need to check if current thread is the main thread. I need to support both console and WPF applications - so I cannot check [SynchronizationContext.Current](http://msdn.microsoft.com/en-us/library/system.threading.synchronizationcontext.current.aspx) or [Application.Current.Dispatcher.Thread](https://learn.microsoft.com/en-us/dotnet/api/system.windows.threading.dispatcher). An I need to support both .NFW 472 and .NET 8.0.
1) Is it reliable to suppose that for a main thread `Thread.CurrentThread.ManagedThreadId` always equals 1?
2) If not - is it enough to check via 
```
if (Thread.CurrentThread.GetApartmentState() == ApartmentState.STA && Thread.CurrentThread.ManagedThreadId == 1 &&
    !Thread.CurrentThread.IsBackground && !Thread.CurrentThread.IsThreadPoolThread && Thread.CurrentThread.IsAlive)
{
    // it is the main thread
}
```
or is it an overkill and there is a simplier solution?
3) Or maybe some extra checks via reflection are needed:
```
public static bool CheckForMainThread()
{
    if (Thread.CurrentThread.GetApartmentState() == ApartmentState.STA && Thread.CurrentThread.ManagedThreadId == 1 &&
        !Thread.CurrentThread.IsBackground && !Thread.CurrentThread.IsThreadPoolThread && Thread.CurrentThread.IsAlive)
    {
        MethodInfo correctEntryMethod = Assembly.GetEntryAssembly().EntryPoint;
        StackTrace trace = new StackTrace();
        StackFrame[] frames = trace.GetFrames();
        for (int i = frames.Length - 1; i >= 0; i--)
        {
            MethodBase method = frames[i].GetMethod();
            if (correctEntryMethod == method)
            {
                return true;
            }
        }
    }
    return false;
}
```
Thank you.

**Comment by jkotas** [comment]
What do you need this check for? In other words, what is your definition of main thread? For example, what's the main thread if the .exe is native binary and the .NET runtime is used to run a COM component?

Neither of the code snippets that you have shared above look reliable.

The fully reliable way would be to have the caller of your code to set the main thread explicitly.

**Reply by alexey-leonovich** [reply]
I need to support UI (WPF, WinForms, Avalonia) and console applications targeting both .NFW 472 and .NET8.0+. Inside them I need to catch exceptions that occur in not-main threads (in operations started with `Task.Run()`) and rethrow them to main thread. 
Now I'm using a common approach with capturing main thread context with:
```
static void CaptureMainContext()
{
    if (_mainThreadContext == null)
    {
        if (SynchronizationContext.Current == null)
            SynchronizationContext.SetSynchronizationContext(new SynchronizationContext());
        _mainThreadContext = SynchronizationContext.Current;
    }
}
```
and adding a continuation to all operations started with `Task.Run()` for rethrowing exception to main thread:
```
public static void RethrowExceptionToMainThread(Task task)
{
    task.ContinueWith(
        tsk =>
        {
            var exceptionDispatchInfo = ExceptionDispatchInfo.Capture(tsk.Exception.InnerException);
            _mainThreadContext.Post(ex => (ex as ExceptionDispatchInfo)?.Throw(), exceptionDispatchInfo);
        }, TaskContinuationOptions.OnlyOnFaulted);
}
```
So I need to ensure that `CaptureMainContext()` is always called on main thread. Handy way to archieve that would be calling inside [IHostApplicationLifetime](https://learn.microsoft.com/en-us/dotnet/core/extensions/generic-host?tabs=appbuilder#ihostapplicationlifetime) `ApplicationStarted` event but that approach is only for an application on .NET Generic Host (`host.Start()` [should be called](https://stackoverflow.com/questions/79219594/correct-usage-of-net-generic-host-and-ihostapplicationlifetime-in-net-8-0-wpf)). So I ended up with this question (howto ensure that current thread is the main thread).
Is there a better way to achieve my goal?

**Reply by jkotas** [reply]
Are you designing a reusable library? Why does the library need to post the exceptions to the main thread?

Reusable libraries typically depend on the synchronization context created by the caller. It is an anti-pattern for reusable library to set the synchronization context.

**Reply by MichalPetryka** [reply]
I agree that `main thread` is probably not the thing you really want here. Any UI framework can very much so not be running on the main thread of the app that started with the entrypoint, in fact such thread might not even ever executed any managed code at all in cases of custom runtime hosting.

What you really want here is to get the thread used by the UI framework and as such, it should be them that provide an API for retrieving it.

**Reply by alexey-leonovich** [reply]
@jkotas Yes, I'm designing a library that I want to reuse in UI (WPF, WinForms, Avalonia) and console applications targeting both .NFW 472 and .NET8.0+. 
In UI applications we check if current thread is main thread via [Application.Current.Dispatcher.CheckAccess()](https://learn.microsoft.com/en-us/dotnet/api/system.windows.threading.dispatcher.checkaccess). And we use it like that:
```
public void ExecuteWithProgressBar(Action actionProcess, string message)
{
    if (Application.Current.Dispatcher.CheckAccess())
        //show progressbar in UI
        ViewManager.ExecuteWithProgressBar(actionProcess, message);
    else
        //perform some operation without UI
        actionProcess();
}
```
In console applications we use [System.Threading.SynchronizationContext)](https://learn.microsoft.com/en-us/dotnet/api/system.threading.synchronizationcontext) to catch exceptions in non-main threads like that:
```
namespace TestConsoleApp
{
    static class Program
    {
        static void Main(string[] args)
        {
            AppDomain.CurrentDomain.UnhandledException += (_, e) => { Console.WriteLine($"{e.ExceptionObject}"); };

            var host = Host.CreateDefaultBuilder(args).Build();
            host.Start();

            if (SynchronizationContext.Current == null)
            {
                SynchronizationContext.SetSynchronizationContext(new SynchronizationContext());
            }
            var context = SynchronizationContext.Current;

            Task.Run(() => throw new Exception("From task");)
                .ContinueWith( a =>
                    {
                        //Rethrow exception from task to main thread
                        context.Post(_ => { throw a.Exception.Flatten(); }, null);
                    }, TaskContinuationOptions.OnlyOnFaulted);
            host.WaitForShutdown();
        }
    }
}
```
And now I need to implement both functionalities in a .netstandard20 reusable library. That's why I need to check that current thread is main thread (without using UI/Windows specific classes).

**Reply by OwnageIsMagic** [reply]
@alexey-leonovich "Main thread" definition is tied to UI framework. Nothing stops you from `Task.Run(() => Application.Run())` to run WinForms on background thread pool thread.  
Btw, your console applications example is bogus. It catches exception from one thread pool thread and just rethrows it on another thread pool thread. See [`SynchronizationContext.Post`](https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Threading/SynchronizationContext.cs,25).

**Reply by huoyaoyuan** [reply]
By definition, there's no common way to check "main thread" defined by different UI framework. There are even special cases like UWP, where multiple AppViews with multiple UI threads can run in the same process.
